### PR TITLE
[shared_preferences_windows]-Migrate to null safety

### PR DESCRIPTION
**Pull request by Sameerkash**
Migrate `shared_preferences_windows` to null safety

Related to issue  flutter/flutter#74876.

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart [hunk 1]
```diff
@@ -21,18 +21,19 @@ void main() {
   tearDown(() {});
 
   Future<String> _getFilePath() async {
-    final directory = await pathProvider.getApplicationSupportPath();
+    final directory =
+        await (pathProvider!.getApplicationSupportPath() as Future<String>);
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart:25]: Please use `!`, not a cast, when you are removing nullability, so it's very clear what the purpose is.

(This isn't technically safe, but since the test will fail if it somehow happens that should be fine.)

File: packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart [hunk 2]
```diff
@@ -87,23 +88,23 @@ void main() {
 /// path it returns is a root path that does not actually exist on Windows.
 class FakePathProviderWindows extends PathProviderPlatform
     implements PathProviderWindows {
-  VersionInfoQuerier versionInfoQuerier;
+  late VersionInfoQuerier versionInfoQuerier;
 
   @override
   Future<String> getApplicationSupportPath() async => r'C:\appsupport';
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart:94]: This is nullable in the interface, so it would be clearer to declare it that way for the override.

File: script/nnbd_plugins.sh [hunk 3]
```diff
@@ -22,6 +22,7 @@ readonly NNBD_PLUGINS_LIST=(
   "url_launcher"
   "video_player"
   "webview_flutter"
+  "shared_preferences_windows"
```
Comment (stuartmorgan) [inline script/nnbd_plugins.sh:25]: I wouldn't expect this to be necessary; everything else here is app-facing. Did you do this to resolve a CI failure?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Thanks, looks like a good start! It's failing tests; I flagged what I expect is the source of that below.

In general, please make sure that integration tests pass locally before submitting PRs.

File: script/nnbd_plugins.sh (hunk 3 above)
Comment (Sameerkash) [reply script/nnbd_plugins.sh:25]: Yes, had a previous PR failing for this reason. So added this.

**Comment by Sameerkash** [comment]
@stuartmorgan  Thanks for reviewing, I'll send push the changes soon. Also, I was not able to format I properly using `flutter_plugin_tools`, it keeps crashing at certain packages. Any ideas?

File: packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart (hunk 1 above)
Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart:25]: Yes, didn't double-check all the casts, after we discussed about adding `!`. Will be careful next time.

File: packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart [hunk 4]
```diff
@@ -20,49 +20,51 @@ class SharedPreferencesWindows extends SharedPreferencesStorePlatform {
...
-    if (_localDataFilePath == null) {
-      final directory = await pathProvider.getApplicationSupportPath();
-      _localDataFilePath =
-          fs.file(path.join(directory, 'shared_preferences.json'));
+    if (_localDataFilePath != null) {
+      return _localDataFilePath!;
     }
-    return _localDataFilePath;
+    final directory =
+        await (pathProvider!.getApplicationSupportPath() as FutureOr<String>);
```
[the remark this answers is not in the document]
Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart:41]: `getApplicationSupportPath()`  returns a Future<String?>

File: script/nnbd_plugins.sh (hunk 3 above)
Comment (gaetschwartz) [reply script/nnbd_plugins.sh:25]: I don't think it is necessary or needed to add it here. You just have a file not complying with the expected format.

File: packages/shared_preferences/shared_preferences_windows/pubspec.yaml [hunk 5]
```diff
@@ -11,20 +12,22 @@ flutter:
...
   flutter: ">=1.12.8"
 
 dependencies:
-  shared_preferences_platform_interface: ^1.0.0
+  shared_preferences_platform_interface: ^2.0.0-nullsafety
   flutter:
     sdk: flutter
-  file: ">=5.1.0 <7.0.0"
+
+  ffi: ^0.2.0-nullsafety.1
```
[the remark this answers is not in the document]
Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/pubspec.yaml:23]: Not sure if I understand which one are you referring to?

File: script/nnbd_plugins.sh (hunk 3 above)
Comment (Sameerkash) [reply script/nnbd_plugins.sh:25]: @gaetschwartz, which one is not complying?

File: packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart (hunk 2 above)
Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart:94]: you mean `getApplicationSupportPath()` ?

File: script/nnbd_plugins.sh (hunk 3 above)
Comment (gaetschwartz) [reply script/nnbd_plugins.sh:25]: Read the checks.  It says :

> These files are not formatted correctly (see diff below):
  shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart

**Comment by gaetschwartz** [comment]
@stuartmorgan I made a PR that migrated this plus `shared_preferences_web` & `shared_preferences_linux` (https://github.com/flutter/plugins/pull/3523). All tests pass locally. Should I revert my migration of `shared_prefrences_windows` or  is my PR okay ? 

@Sameerkash Didn't mention the main issue so I missed this PR and ported all the platform plugins.

**Comment by stuartmorgan** [comment]
> @stuartmorgan I made a PR that migrates this plus `shared_preferences_web` & `shared_preferences_linux` (#3523). All tests pass locally. Should I revert my migration of `shared_prefrences_windows` or is my PR okay ?

Just commented there; smaller PRs are generally better, so I would rather proceed with separate PRs. Thanks!

**Comment by stuartmorgan** [comment]
> Also, I was not able to format I properly using `flutter_plugin_tools`, it keeps crashing at certain packages. Any ideas?

Could you file an issue about this with details?

File: packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart [hunk 6]
```diff
@@ -27,35 +27,37 @@ class SharedPreferencesWindows extends SharedPreferencesStorePlatform {
...
-      final directory = await pathProvider.getApplicationSupportPath();
-      _localDataFilePath =
-          fs.file(path.join(directory, 'shared_preferences.json'));
+    if (_localDataFilePath != null) {
+      return _localDataFilePath!;
     }
-    return _localDataFilePath;
+    final directory = await pathProvider.getApplicationSupportPath();
+    return _localDataFilePath =
+        fs.file(path.join(directory!, 'shared_preferences.json'));
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart:42]: Missed this. This isn't safe; `directory` is nullable because `getApplicationSupportPath` returns a nullable value. This needs to be checked, and a null return value propagated and handled.

(For read we can return a empty dictionary as with a missing file; for write I think we'll need to throw an error, as we don't know where we can write to in that case.)

File: packages/shared_preferences/shared_preferences_windows/pubspec.yaml [hunk 7]
```diff
@@ -11,20 +12,21 @@ flutter:
         pluginClass: none
 
 environment:
-  sdk: ">=2.1.0 <3.0.0"
+  sdk: '>=2.12.0-259.8.beta <3.0.0'
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/pubspec.yaml:15]: Let's stay consistent about using `"`

Why the specific version here, rather than `-0`?

File: packages/shared_preferences/shared_preferences_windows/pubspec.yaml [hunk 8]
```diff
@@ -11,20 +12,21 @@ flutter:
...
-  path_provider_platform_interface: ^1.0.3
-  path_provider_windows: ^0.0.2
+  path_provider_platform_interface: ^2.0.0-nullsafety
+  path_provider_windows: ^0.1.0-nullsafety.2
 
 dev_dependencies:
   flutter_test:
     sdk: flutter
-  pedantic: ^1.8.0
+  pedantic: ^1.10.0-nullsafety.3
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/pubspec.yaml:32]: Please fix the missing newline.

File: packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart [hunk 9]
```diff
@@ -46,7 +46,7 @@ void main() {
     await _writeTestFile('{"key1": "one", "key2": 2}');
     var prefs = _getPreferences();
 
-    var values = await prefs.getAll();
+    var values = await (prefs.getAll());
```
Comment (stuartmorgan) [inline packages/shared_preferences/shared_preferences_windows/test/shared_preferences_windows_test.dart:49]: Why the ()?

File: script/nnbd_plugins.sh [hunk 10]
```diff
@@ -42,4 +42,4 @@ readonly NON_NNBD_PLUGINS_LIST=(
 )
 
 export EXCLUDED_PLUGINS_FROM_STABLE=$(IFS=, ; echo "${NNBD_PLUGINS_LIST[*]}")
-export EXCLUDED_PLUGINS_FROM_MASTER=$(IFS=, ; echo "${NON_NNBD_PLUGINS_LIST[*]}")
+export EXCLUDED_PLUGINS_FROM_MASTER=$(IFS=, ; echo "${NON_NNBD_PLUGINS_LIST[*]}")
```
Comment (stuartmorgan) [inline script/nnbd_plugins.sh:45]: Please revert this change.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Looking good, just a few small things. Thanks!

File: packages/shared_preferences/shared_preferences_windows/pubspec.yaml (hunk 7 above)
Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/pubspec.yaml:15]: I believe pub was not resolving at all, hence changed the version here.

Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/pubspec.yaml:15]: I'll try changing and see if ti throws any errors.

File: packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart (hunk 6 above)
Comment (Sameerkash) [reply packages/shared_preferences/shared_preferences_windows/lib/shared_preferences_windows.dart:42]: could you give me a bit more pointers on how the code should be?  wrap it in a `try` and  `catch`?

File: script/nnbd_plugins.sh (hunk 10 above)
Comment (Sameerkash) [reply script/nnbd_plugins.sh:45]: The new line?

**Review by stuartmorgan — APPROVED** [review]
I went ahead and did the last fixes, since I'd like to try to get this through CI, landed, and published this morning if possible to unblock shared_preferences in general.

**Comment by Sameerkash** [comment]
> I went ahead and did the last fixes, since I'd like to try to get this through CI, landed, and published this morning if possible to unblock shared_preferences in general.

I had made all changes except in `_writepreferences()` but sure!. 

I have another package in migrating that I need shared preferences and yes it'd be good if it's published soon. 


And thanks for all the guidance. I learnt a lot during this PR as it required a lot of manaul refactoring.
